### PR TITLE
retire table validation from apex

### PR DESCRIPTION
**Pull request by RetYn**
retire table validation in Apex and migrate to Docs
[AB#372101](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/372101)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/docfx/pull/7076)

### CODE REVIEW

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 1]
```diff
@@ -231,5 +240,126 @@ static string GetContainerInlineContent(ContainerInline containerInline)
...
+        {
+            if (cell == null || cell.Count == 0)
+            {
+                return new TableCellNode
+                {
+                    IsVisible = false,
+                    IsEmphasis = false,
+                };
+            }
+            var block = cell.LastChild;
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:315]: why use the Last Child? consider below case:
```md
| asf | fda | 
| -- | -- |
| sdafasd ![sdf](asdfdasf) **sadffd** sds | asdfdsf |

```

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 2]
```diff
@@ -231,5 +240,126 @@ static string GetContainerInlineContent(ContainerInline containerInline)
...
+            }
+
+            var plainText = markdownEngine.ToPlainText(paragraph);
+
+            var lines = plainText.Split("\n").ToList();
+            bool tableHeaderExist = false;
+            int tableDelimiterExistLine = 0;
+            foreach (var line in lines)
+            {
+                if (line.Contains("|"))
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:349]: "-" line should follow up with the line contains "|", and we can break earlier.

var tableDelimiterExistsLine = 0;
foreach(var line in lines)
{
var cells = line.Split("|");
if (!cells.Any())
{
    tableDelimiterExistsLine = 0;
    continue;
}

if (tableDelimiterExistsLine ++ > 0 && cells.Any(c => c.contains ("-")))
    return true;
}

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 3]
```diff
@@ -231,5 +240,124 @@ static string GetContainerInlineContent(ContainerInline containerInline)
...
+            }
+        }
+
+        private static TableNode TryParseTable(Table table)
+        {
+            if (table.Count == 0)
+            {
+                return new TableNode { IsSuccessParsed = false, };
+            }
+            var columnHeaders = new List<TableCellNode>();
```
Comment (yufeih) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:279]: Consider turn `TableCellNode` into a struct to reduce memory allocation.

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs:315]: We could say in order to be an emphasis, aside from whitespaces, the block MUST be a paragraph block that starts and ends with `EmphasisInline`.

Is `*a*` also considered valid or we only allow `**a**`? `EmphasisInline.DelimiterCount` tells the difference.

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 4]
```diff
@@ -231,5 +240,124 @@ static string GetContainerInlineContent(ContainerInline containerInline)
...
+                case Table table:
+                    var parsedNode = TryParseTable(table);
+                    tableNode = CreateValidationNode<TableNode>(isCanonicalVersion, node) with
+                    {
+                        RowHeaders = parsedNode.RowHeaders,
+                        ColumnHeaders = parsedNode.ColumnHeaders,
+                        IsSuccessParsed = parsedNode.IsSuccessParsed,
+                    };
+                    break;
+                case ParagraphBlock:
```
Comment (yufeih) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:258]: Inline type cast:
```suggestion
                case ParagraphBlock paragraph:
```

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 5]
```diff
@@ -231,5 +240,124 @@ static string GetContainerInlineContent(ContainerInline containerInline)
...
+        }
+
+        private static bool TryDetectTable(ParagraphBlock paragraph, MarkdownEngine markdownEngine)
+        {
+            if (!paragraph.Inline.FindDescendants<LineBreakInline>().Any())
+            {
+                return false;
+            }
+
+            var plainText = markdownEngine.ToPlainText(paragraph);
```
Comment (yufeih) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:342]: `ToPlainText` for every paragraph with a `LineBreakInline` is a pretty expensive operation. Consider traverse the `ParagraphBlock` DOM to determine if this is likely a misconfigured table.

Performance Tips: minimize allocation for read operations.

**Review by yufeih — CHANGES_REQUESTED** [review]
Need regression test to check for correctness and performance.

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 6]
```diff
@@ -240,14 +240,14 @@ public static bool IsVisible(this MarkdownObject markdownObject)
                     ThematicBreakBlock _ => false,
                     YamlFrontMatterBlock _ => false,
                     HeadingBlock headingBlock when headingBlock.Inline is null || !headingBlock.Inline.Any() => false,
-                    LeafBlock leafBlock when leafBlock.Inline is null || !leafBlock.Inline.Any() => true,
+                    CodeBlock codeBlock when codeBlock.Lines.Count != 0 => true,
+                    LeafBlock leafBlock when leafBlock.Inline is null || !leafBlock.Inline.Any() => false,
                     LinkInline linkInline when linkInline.IsImage => true,
                     TripleColonInline tripleColonInline when tripleColonInline.Extension is ImageExtension => true,
-                    LiteralInline literal when !string.IsNullOrWhiteSpace(literal.Content.Text) => true,
+                    LiteralInline literal when string.IsNullOrWhiteSpace(literal.Content.ToString()) => false,
```
[the remark this answers is not in the document]
Comment (RetYn) [reply src/docfx/lib/markdown/MarkdigUtility.cs:247]: Context.Text contains "\0".. and string.IsNullOrWhiteSpace will be true

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 7]
```diff
@@ -277,7 +277,7 @@ private static bool IsInlineImage(this Inline node)
                         {
                             foreach (var child in containerInline)
                             {
-                                if (child != current && child.IsVisible())
+                                if (child.IsVisible())
```
[the remark this answers is not in the document]
Comment (RetYn) [reply src/docfx/lib/markdown/MarkdigUtility.cs:280]: It old logic is uncorrect. It ignore the node itself。

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 8]
```diff
@@ -231,5 +240,124 @@ static string GetContainerInlineContent(ContainerInline containerInline)
...
+            {
+                tableNodes.Add(tableNode);
+            }
+        }
+
+        private static TableNode TryParseTable(Table table)
+        {
+            if (table.Count == 0)
+            {
+                return new TableNode { IsSuccessParsed = false, };
```
[the remark this answers is not in the document]
Comment (RetYn) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs:277]: Do you mean that we trust markdig and do not deal some null or zero cases？

**Review by yufeih — APPROVED** [review]

**Pull request merged**